benchmarking/generate_l1_svm_lp.jl

```diff
@@ -0,0 +1,182 @@
+#     --output_filename=/tmp/duke.mps.gz --regularizer_weight=1.5
+
+# The LP formulation for L1 SVM is taken from equation (5) in:
+# Ji Zhu, Saharon Rosset, Trevor Hastie, and Rob Tibshirani. 2003. 1-norm
+# support vector machines. In Proceedings of the 16th International Conference
+# on Neural Information Processing Systems (NIPS'03). MIT Press, Cambridge, MA,
+# USA, 49–56.
+# https://papers.nips.cc/paper/2003/file/49d4b2faeb4b7b9e745775793141e2b2-Paper.pdf
+
+import ArgParse
+import JuMP
+import SparseArrays
+import LinearAlgebra
+
+const SparseMatrixCSC = SparseArrays.SparseMatrixCSC
+const sparse = SparseArrays.sparse
+const norm = LinearAlgebra.norm
+const Diagonal = LinearAlgebra.Diagonal
+
+include("utils.jl")
+
+mutable struct SvmTrainigData
+  feature_matrix::SparseMatrixCSC{Float64,Int64}
+  labels::Vector{Float64}
+end
+
+"""
+Populates 'model' with the linear program model for solving the L1 SVM
+problem with training set (X, y). The model is given by
+  min  sum w[i] + regularizer_weight * sum z[i]
+  s.t. w >= 1 - diag(y) * [X 1] * beta
+       w >= 0
+       z >= beta
+       z >= -beta
+"""
+function populate_libsvm_model(
+  model::JuMP.Model,
+  data::SvmTrainigData,
+  regularizer_weight::Float64,
+)
+  n, d = size(data.feature_matrix)
+  println("Generating a model with $n datapoints and $(d - 1) features.")
+  JuMP.@variable(model, beta[i = 1:d])
+  JuMP.@variable(model, w[i = 1:n], lower_bound = 0.0)
+  JuMP.@variable(model, z[i = 1:d])
+  JuMP.@objective(model, Min, sum(w) + regularizer_weight * sum(z))
+  JuMP.@constraint(model, z .>= beta)
+  JuMP.@constraint(model, z .>= -beta)
+  JuMP.@constraint(
+    model,
+    w .>= (1.0 .- data.labels .* (data.feature_matrix * beta))
+  )
+  return model
+end
+
+"""
+Defines parses and args.
+
+# Returns
+A dictionary with the values of the command-line arguments.
+"""
+function parse_command_line()
+  arg_parse = ArgParse.ArgParseSettings()
+
+  ArgParse.@add_arg_table! arg_parse begin
+    "--input_filename"
+    help = "The .train file containing the problem data."
+    arg_type = String
+    required = true
+
+    "--output_filename"
+    help = "Filename for the output .mps (or .mps.gz) model."
+    arg_type = String
+    required = true
+
+    "--regularizer_weight"
+    help = "Weight of the L1 regularizer."
+    arg_type = Float64
+    required = true
+  end
+
+  return ArgParse.parse_args(arg_parse)
+end
+
+"""
+Loads a LIBSVM file and returns a loaded SvmTrainigData struct.
+"""
+function load_libsvm_file(file_name::String)
+  open(file_name, "r") do io
+    labels = Vector{Float64}()
+    row_indices = Vector{Int64}()
+    col_indices = Vector{Int64}()
+    matrix_values = Vector{Float64}()
+
+    row_index = 0
+    for line in eachline(io)
+      row_index += 1
+      split_line = split(line)
+
+      label = parse(Float64, split_line[1])
+      # This ensures that labels are 1 or -1. Different datasets use {-1, 1}, {0, 1}, and {1, 2}.
+      if label ≈ 1.0
```

Comment:
What's the intention of the approximate equality? It doesn't include much:
```julia
julia> 1.0001 ≈ 1
false
```

Reply:
I changed it and added an assert that checks if a label one was found.